File: AlphaZ-Core/src/main/java/com/alphaz/core/pojo/entity/BaseDO.java
package com.alphaz.core.pojo.entity;

import com.alphaz.core.constant.DataState;
import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.*;
import java.sql.Timestamp;


/**
 * 持久化对象的基类,所有要持久化到数据库的对象继承该类或其子类
 *
 * @author HuangJian
 * @since May 19, 2010
 */
@MappedSuperclass
public abstract class BaseDO {

    /**
     * 主键，UUID
     */
    private Long id;

    /**
     * 逻辑删除
     */
    private DataState state;
    /**
     * 创建时间
     */
    private Timestamp createTime;
    /**
     * 更新时间
     */
    private Timestamp updateTime;
    /**
     * 创建人
     */
    private Long createBy;
    /**
     * 更新人
     */
    private Long updateBy;

    public void setId(Long id) {
        this.id = id;
    }

    @Basic
    @Column(name = "state", nullable = false, columnDefinition = "tinyint default 1", length = 1)
    public DataState getState() {
        return state;
    }

    public void setState(DataState state) {
        this.state = state;
    }

    @Basic
    @Column(name = "create_time", columnDefinition = "datetime default CURRENT_TIMESTAMP")
    public Timestamp getCreateTime() {
        return createTime;
    }

    public void setCreateTime(Timestamp createTime) {
        this.createTime = createTime;
    }

    @Basic
    @Column(name = "update_time", columnDefinition = "datetime on update CURRENT_TIMESTAMP")
    public Timestamp getUpdateTime() {
        return updateTime;
    }

    public void setUpdateTime(Timestamp updateTime) {
        this.updateTime = updateTime;
    }
    @Basic
    public Long getCreateBy() {
        return createBy;
    }

    public void setCreateBy(Long createBy) {
        this.createBy = createBy;
    }
    @Basic
    public Long getUpdateBy() {
        return updateBy;
    }

    public void setUpdateBy(Long updateBy) {
        this.updateBy = updateBy;
    }

    @Version
    protected Integer version;

    @Id
    @Column(name = "id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long getId() {
        return id;
    }

    @Basic
    @Column(name = "version", nullable = false, columnDefinition = "int default 1", length = 1)
    public Integer getVersion() {
        return version;
    }

    public void setVersion(Integer version) {
        this.version = version;
    }
}

File: Infrastructure/src/main/kotlin/com/alphaz/infrastructure/domain/BaseDO.kt
package com.alphaz.infrastructure.domain

import com.alphaz.infrastructure.constant.State
import com.alphaz.infrastructure.exception.NotAValueObjectException
import com.alphaz.infrastructure.exception.NotAnEntityException
import org.hibernate.annotations.Filter
import org.hibernate.annotations.FilterDef
import org.hibernate.annotations.ParamDef
import java.time.LocalDateTime
import javax.persistence.*
import kotlin.jvm.Transient
import javax.persistence.PrePersist


/**
 *@Author: c0der
 *@Date: 下午4:30 2018/3/15
 *@Description:An entity, as explained in the DDD，对于entity的独有操作需要在entity中实现
 */
@MappedSuperclass
@FilterDef(name = "dataStateCondition", parameters = arrayOf(ParamDef(name = "state", type = "int")))
@Filter(name = "dataStateCondition", condition = "state = :state")
abstract class BaseDO<C : DO<*, *>, T> : DO<C, T> {
    override fun equal(other: C): Boolean {
        return this.id == other.id;
    }

    /**
     * 主键，UUID
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    override var id: T? = null

    /**
     * 逻辑删除
     */
    @Column(name = "state", nullable = true, columnDefinition = "tinyint default 1", length = 1)
    override var state: State = State.ACTIVE
    /**
     * 创建时间
     */
    @Column(name = "create_time", columnDefinition = "datetime default CURRENT_TIMESTAMP")
    override var createTime: LocalDateTime = LocalDateTime.now()
    /**
     * 更新时间
     */
    @Column(name = "update_time", columnDefinition = "datetime on update CURRENT_TIMESTAMP")
    override var updateTime: LocalDateTime? = null

    /**
     * 创建人
     */
    override var createBy: T? = null

    /**
     * 更新人
     */
    override var updateBy: T? = null

    override var owner: T? = null

    @Column(name = "version", nullable = false, columnDefinition = "int default 1", length = 1)
    override var version: Int = 0

    @Version
    @Transient
    var lock: Int? = null


}